filebeat/input/log/input.go

```diff
@@ -727,14 +731,29 @@ func (p *Input) Wait() {
 
 // Stop stops all harvesters and then stops the input
 func (p *Input) Stop() {
-	// Stop all harvesters
-	// In case the beatDone channel is closed, this will not wait for completion
-	// Otherwise Stop will wait until output is complete
-	p.harvesters.Stop()
+	p.stopOnce.Do(func() {
+		// Stop all harvesters
+		// In case the beatDone channel is closed, this will not wait for completion
+		// Otherwise Stop will wait until output is complete
+		p.harvesters.Stop()
+
+		// close state updater
+		p.stateOutlet.Close()
+
+		// stop all communication between harvesters and publisher pipeline
+		p.outlet.Close()
+	})
+}
 
-	// close state updater
-	p.stateOutlet.Close()
+// stopWhenDone takes care of stopping the input if some of the contexts are done
+func (p *Input) stopWhenDone() {
+	go func() {
```

Comment:
perhaps making this a common function and doing `go p.stopWhenDone()` would be more explicit? I don't have an strong opinion anyway